src/ripple/app/ledger/Ledger.cpp

```diff
@@ -599,6 +605,119 @@ Ledger::peek(Keylet const& k) const
+    {
+        if (sle->isFieldPresent(sfNegativeUNL))
+        {
+            auto const& nUnlData = sle->getFieldArray(sfNegativeUNL);
+            for (auto const& n : nUnlData)
+            {
+                if (n.isFieldPresent(sfPublicKey))
+                {
+                    auto d = n.getFieldVL(sfPublicKey);
+                    auto s = makeSlice(d);
+                    if (!publicKeyType(s))
+                    {
+                        continue;
+                    }
+                    nUnl.emplace(s);
+                }
+            }
+        }
+    }
+
+    return nUnl;
+}
+
+boost::optional<PublicKey>
+Ledger::nUnlToDisable() const
+{
+    if (auto sle = read(keylet::negativeUNL()))
+    {
+        if (sle->isFieldPresent(sfNegativeUNLToDisable))
+        {
+            auto d = sle->getFieldVL(sfNegativeUNLToDisable);
+            auto s = makeSlice(d);
+            if (publicKeyType(s))
+                return PublicKey(s);
+        }
+    }
+
+    return boost::none;
+}
+
+boost::optional<PublicKey>
+Ledger::nUnlToReEnable() const
+{
+    if (auto sle = read(keylet::negativeUNL()))
+    {
+        if (sle->isFieldPresent(sfNegativeUNLToReEnable))
+        {
+            auto d = sle->getFieldVL(sfNegativeUNLToReEnable);
+            auto s = makeSlice(d);
+            if (publicKeyType(s))
+                return PublicKey(s);
+        }
+    }
+
+    return boost::none;
+}
+
+void
+Ledger::updateNegativeUNL()
+{
+    if (info_.seq % 256 == 0)
+    {
+        if (auto sle = peek(keylet::negativeUNL()))
+        {
+            bool hasToDisable = sle->isFieldPresent(sfNegativeUNLToDisable);
+            bool hasToReEnable = sle->isFieldPresent(sfNegativeUNLToReEnable);
+            if (hasToDisable || hasToReEnable)
+            {
+                STArray newNUnl;
+                if (sle->isFieldPresent(sfNegativeUNL))
+                {
+                    auto const& oldNUnl = sle->getFieldArray(sfNegativeUNL);
+                    for (auto v : oldNUnl)
+                    {
+                        if (hasToReEnable && v.isFieldPresent(sfPublicKey) &&
+                            v.getFieldVL(sfPublicKey) ==
+                                sle->getFieldVL(sfNegativeUNLToReEnable))
+                            continue;
+                        newNUnl.push_back(v);
+                    }
+                }
+
+                if (hasToDisable)
+                {
+                    newNUnl.emplace_back(sfNegativeUNLEntry);
+                    newNUnl.back().setFieldVL(
+                        sfPublicKey, sle->getFieldVL(sfNegativeUNLToDisable));
+                    newNUnl.back().setFieldU32(sfFirstLedgerSequence, seq());
+                }
+
+                if (!newNUnl.empty())
+                {
+                    sle->setFieldArray(sfNegativeUNL, newNUnl);
+                    if (hasToReEnable)
+                        sle->delField(sfNegativeUNLToReEnable);
```

Comment:
I _think_ you should use `makeFieldAbsent` instead of `delField`. I think `delField` executed by itself puts the object in an invalid state. `makeFieldAbsent` does what you want. If you look through the codebase, `delField` is only used internally, in `STObject`. `makeFieldAbsent` is used in a lot of places.

Reply:
Thanks so much!